File: gateway/node-service/src/routes/weather.ts
import { Router } from 'express';
import axios from 'axios';

export const weatherRouter = Router();

const WEATHER_API_BASE = 'https://api.weather.gov';
const USER_AGENT = 'AcreBlitz Gateway (https://acreblitz.com)';

// Cache for grid point data
const gridPointCache = new Map<
  string,
  { data: GridPointData; timestamp: number }
>();
const GRID_CACHE_TTL = 24 * 60 * 60 * 1000; // 24 hours

interface GridPointData {
  gridId: string;
  gridX: number;
  gridY: number;
  forecastHourly: string;
  observationStations: string;
  city: string;
  state: string;
}

interface WeatherData {
  location: {
    city: string;
    state: string;
    gridId: string;
    gridX: number;
    gridY: number;
  };
  currentConditions: {
    timestamp: string;
    temperature: number | null;
    temperatureUnit: string;
    description: string;
    icon: string;
    humidity: number | null;
    windSpeed: number | null;
    windDirection: number | null;
    pressure: number | null;
  } | null;
  hourlyForecast: Array<{
    time: string;
    temperature: number;
    temperatureUnit: string;
    precipitationChance: number | null;
    relativeHumidity: number | null;
    windSpeed: string;
    windDirection: string;
    icon: string;
    shortForecast: string;
    isDaytime: boolean;
  }>;
  updated: string;
}

function celsiusToFahrenheit(celsius: number | null): number | null {
  if (celsius === null) return null;
  return Math.round((celsius * 9) / 5 + 32);
}

async function getGridPoint(lat: number, lon: number): Promise<GridPointData> {
  const cacheKey = `${lat.toFixed(4)},${lon.toFixed(4)}`;
  const cached = gridPointCache.get(cacheKey);

  if (cached && Date.now() - cached.timestamp < GRID_CACHE_TTL) {
    console.log('[Weather] Using cached grid point for:', cacheKey);
    return cached.data;
  }

  console.log('[Weather] Fetching grid point for:', cacheKey);

  const response = await axios.get(`${WEATHER_API_BASE}/points/${lat},${lon}`, {
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'application/geo+json',
    },
    timeout: 10000,
  });

  const props = response.data.properties;
  const data: GridPointData = {
    gridId: props.gridId,
    gridX: props.gridX,
    gridY: props.gridY,
    forecastHourly: props.forecastHourly,
    observationStations: props.observationStations,
    city: props.relativeLocation.properties.city,
    state: props.relativeLocation.properties.state,
  };

  gridPointCache.set(cacheKey, { data, timestamp: Date.now() });
  return data;
}

async function getHourlyForecast(url: string) {
  console.log('[Weather] Fetching hourly forecast');

  const response = await axios.get(url, {
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'application/geo+json',
    },
    timeout: 10000,
  });

  return response.data.properties.periods.map((period: any) => ({
    time: period.startTime,
    temperature: period.temperature,
    temperatureUnit: period.temperatureUnit,
    precipitationChance: period.probabilityOfPrecipitation?.value ?? null,
    relativeHumidity: period.relativeHumidity?.value ?? null,
    windSpeed: period.windSpeed,
    windDirection: period.windDirection,
    icon: period.icon,
    shortForecast: period.shortForecast,
    isDaytime: period.isDaytime,
  }));
}

async function getCurrentConditions(stationsUrl: string) {
  try {
    console.log('[Weather] Fetching observation stations');

    const stationsResponse = await axios.get(stationsUrl, {
      headers: {
        'User-Agent': USER_AGENT,
        Accept: 'application/geo+json',
      },
      timeout: 10000,
    });

    const stations = stationsResponse.data.features;
    if (!stations || stations.length === 0) {
      console.warn('[Weather] No observation stations found');
      return null;
    }

    const stationId = stations[0].id;
    console.log('[Weather] Fetching latest observation from:', stationId);

    const obsResponse = await axios.get(`${stationId}/observations/latest`, {
      headers: {
        'User-Agent': USER_AGENT,
        Accept: 'application/geo+json',
      },
      timeout: 10000,
    });

    const props = obsResponse.data.properties;
    const tempC = props.temperature?.value;
    const tempF = celsiusToFahrenheit(tempC);

    return {
      timestamp: props.timestamp,
      temperature: tempF,
      temperatureUnit: 'F',
      description: props.textDescription || 'N/A',
      icon: props.icon || '',
      humidity: props.relativeHumidity?.value ?? null,
      windSpeed: props.windSpeed?.value ?? null,
      windDirection: props.windDirection?.value ?? null,
      pressure: props.barometricPressure?.value ?? null,
    };
  } catch (error) {
    console.error('[Weather] Error fetching current conditions:', error);
    return null;
  }
}

weatherRouter.get('/forecast', async (req, res) => {
  try {
    const { lat, lon } = req.query;

    // Validate parameters
    if (!lat || !lon) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'Missing required parameters: lat and lon',
      });
    }

    const latitude = parseFloat(lat as string);
    const longitude = parseFloat(lon as string);

    if (isNaN(latitude) || isNaN(longitude)) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'lat and lon must be valid numbers',
      });
    }

    if (latitude < -90 || latitude > 90) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'lat must be between -90 and 90',
      });
    }

    if (longitude < -180 || longitude > 180) {
      return res.status(400).json({
        error: 'Bad Request',
        message: 'lon must be between -180 and 180',
      });
    }

    console.log('[Weather] Getting weather for:', { latitude, longitude });

    // Get grid point information
    const gridPoint = await getGridPoint(latitude, longitude);

    // Fetch hourly forecast and current conditions in parallel
    const [hourlyForecast, currentConditions] = await Promise.all([
      getHourlyForecast(gridPoint.forecastHourly),
      getCurrentConditions(gridPoint.observationStations),
    ]);

    const weatherData: WeatherData = {
      location: {
        city: gridPoint.city,
        state: gridPoint.state,
        gridId: gridPoint.gridId,
        gridX: gridPoint.gridX,
        gridY: gridPoint.gridY,
      },
      currentConditions,
      hourlyForecast,
      updated: new Date().toISOString(),
    };

    res.json(weatherData);
  } catch (error) {
    console.error('[Weather] Error:', error);

    if (axios.isAxiosError(error)) {
      const status = error.response?.status || 500;
      return res.status(status).json({
        error: 'Weather API Error',
        message: error.message,
        details: error.response?.data,
      });
    }

    res.status(500).json({
      error: 'Internal Server Error',
      message: error instanceof Error ? error.message : 'Unknown error',
    });
  }
});

File: gateway/python-service/routers/weather.py
"""Weather endpoints using National Weather Service API."""

import time
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, HTTPException, Query
import httpx

router = APIRouter(prefix="/weather", tags=["weather"])

WEATHER_API_BASE = "https://api.weather.gov"
USER_AGENT = "AcreBlitz Gateway (https://acreblitz.com)"

# Grid point cache
grid_point_cache: dict[str, dict] = {}
GRID_CACHE_TTL = 24 * 60 * 60  # 24 hours in seconds


def celsius_to_fahrenheit(celsius: Optional[float]) -> Optional[int]:
    """Convert Celsius to Fahrenheit."""
    if celsius is None:
        return None
    return round((celsius * 9) / 5 + 32)


async def get_grid_point(lat: float, lon: float) -> dict:
    """Get grid point information for coordinates."""
    cache_key = f"{lat:.4f},{lon:.4f}"
    cached = grid_point_cache.get(cache_key)

    if cached and time.time() - cached["timestamp"] < GRID_CACHE_TTL:
        print(f"[Weather] Using cached grid point for: {cache_key}")
        return cached["data"]

    print(f"[Weather] Fetching grid point for: {cache_key}")

    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{WEATHER_API_BASE}/points/{lat},{lon}",
            headers={"User-Agent": USER_AGENT, "Accept": "application/geo+json"},
            timeout=10.0,
        )
        response.raise_for_status()
        props = response.json()["properties"]

        data = {
            "gridId": props["gridId"],
            "gridX": props["gridX"],
            "gridY": props["gridY"],
            "forecastHourly": props["forecastHourly"],
            "observationStations": props["observationStations"],
            "city": props["relativeLocation"]["properties"]["city"],
            "state": props["relativeLocation"]["properties"]["state"],
        }

        grid_point_cache[cache_key] = {"data": data, "timestamp": time.time()}
        return data


async def get_hourly_forecast(url: str) -> list:
    """Get hourly forecast from NWS."""
    print("[Weather] Fetching hourly forecast")

    async with httpx.AsyncClient() as client:
        response = await client.get(
            url,
            headers={"User-Agent": USER_AGENT, "Accept": "application/geo+json"},
            timeout=10.0,
        )
        response.raise_for_status()
        periods = response.json()["properties"]["periods"]

        return [
            {
                "time": period["startTime"],
                "temperature": period["temperature"],
                "temperatureUnit": period["temperatureUnit"],
                "precipitationChance": period.get("probabilityOfPrecipitation", {}).get(
                    "value"
                ),
                "relativeHumidity": period.get("relativeHumidity", {}).get("value"),
                "windSpeed": period["windSpeed"],
                "windDirection": period["windDirection"],
                "icon": period["icon"],
                "shortForecast": period["shortForecast"],
                "isDaytime": period["isDaytime"],
            }
            for period in periods
        ]


async def get_current_conditions(stations_url: str) -> Optional[dict]:
    """Get current conditions from nearest observation station."""
    try:
        print("[Weather] Fetching observation stations")

        async with httpx.AsyncClient() as client:
            stations_response = await client.get(
                stations_url,
                headers={"User-Agent": USER_AGENT, "Accept": "application/geo+json"},
                timeout=10.0,
            )
            stations_response.raise_for_status()
            stations = stations_response.json()["features"]

            if not stations:
                print("[Weather] No observation stations found")
                return None

            station_id = stations[0]["id"]
            print(f"[Weather] Fetching latest observation from: {station_id}")

            obs_response = await client.get(
                f"{station_id}/observations/latest",
                headers={"User-Agent": USER_AGENT, "Accept": "application/geo+json"},
                timeout=10.0,
            )
            obs_response.raise_for_status()
            props = obs_response.json()["properties"]

            temp_c = props.get("temperature", {}).get("value")
            temp_f = celsius_to_fahrenheit(temp_c)

            return {
                "timestamp": props.get("timestamp"),
                "temperature": temp_f,
                "temperatureUnit": "F",
                "description": props.get("textDescription") or "N/A",
                "icon": props.get("icon") or "",
                "humidity": props.get("relativeHumidity", {}).get("value"),
                "windSpeed": props.get("windSpeed", {}).get("value"),
                "windDirection": props.get("windDirection", {}).get("value"),
                "pressure": props.get("barometricPressure", {}).get("value"),
            }

    except Exception as e:
        print(f"[Weather] Error fetching current conditions: {e}")
        return None


@router.get("/forecast")
async def get_forecast(
    lat: float = Query(..., ge=-90, le=90, description="Latitude"),
    lon: float = Query(..., ge=-180, le=180, description="Longitude"),
):
    """
    Get current conditions and hourly forecast for a location.

    Uses the National Weather Service API (api.weather.gov).
    Only works for US locations.
    """
    try:
        print(f"[Weather] Getting weather for: lat={lat}, lon={lon}")

        # Get grid point information
        grid_point = await get_grid_point(lat, lon)

        # Fetch hourly forecast and current conditions
        hourly_forecast = await get_hourly_forecast(grid_point["forecastHourly"])
        current_conditions = await get_current_conditions(
            grid_point["observationStations"]
        )

        return {
            "location": {
                "city": grid_point["city"],
                "state": grid_point["state"],
                "gridId": grid_point["gridId"],
                "gridX": grid_point["gridX"],
                "gridY": grid_point["gridY"],
            },
            "currentConditions": current_conditions,
            "hourlyForecast": hourly_forecast,
            "updated": datetime.utcnow().isoformat() + "Z",
        }

    except httpx.HTTPStatusError as e:
        raise HTTPException(
            status_code=e.response.status_code,
            detail=f"Weather API error: {e.response.text}",
        )
    except httpx.RequestError as e:
        raise HTTPException(
            status_code=503,
            detail=f"Weather API unavailable: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch weather: {str(e)}",
        )
